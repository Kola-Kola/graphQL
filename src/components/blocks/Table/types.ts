export interface IBody {
  title: string;
}

export interface IProps {
  body: any;
  withSelectedChoice: boolean;
}
