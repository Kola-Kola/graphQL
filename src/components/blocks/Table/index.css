@media only screen and (max-width: 760px),
  (min-device-width: 768px) and (max-device-width: 1024px) {
  table,
  thead,
  tbody,
  th,
  tr {
    display: block;
  }

  thead tr {
    position: relative;
  }

  tbody tr {
    position: relative;
    min-height: 140px;
  }

  td:first-child {
    position: absolute;
    left: 0;
    top: 0;
    width: 80px;
  }

  td:nth-child(2) {
    display: none;
  }

  td:nth-child(5),
  td:nth-child(4),
  td:nth-child(3) {
    position: absolute;
    left: 50%;
    transform: translateX(-50%);
    max-width: 130px;
    width: 100%;
  }

  td:nth-child(3) {
    top: 20px;
  }

  td:nth-child(4) {
    top: 60px;
    overflow: hidden;
    white-space: nowrap; /* Don't forget this one */
    text-overflow: ellipsis;
  }

  td:nth-child(5) {
    top: 80px;
  }

  td:nth-child(6) {
    position: absolute;
    top: 12px;
    right: 0;
    display: flex;
    align-items: center;
    justify-content: flex-end;
  }

  td {
    border-bottom: 0;
  }

  thead tr th:nth-child(3) {
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translateX(-50%) translateY(-50%);
  }

  thead tr th:not(thead tr th:nth-child(3)) {
    display: none;
  }

  thead tr th:first-child {
    display: block !important;
  }
}
