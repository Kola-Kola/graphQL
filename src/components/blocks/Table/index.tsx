import * as React from "react";
import Checkbox from "../Checkbox";
import { IProps } from "./types";
import { ReactComponent as Information } from "../../assets/information.svg";
import { ReactComponent as Actions } from "../../assets/actions.svg";
import {
  TableContainer,
  TableHeader,
  TableBody,
  TableRow,
  TableHead,
  TableCell,
  TableActionsContainer,
  TableActions,
  TableCheckboxHeaderContainer,
  Row,
} from "./styles";

import "./index.css";

const Table = ({ body, withSelectedChoice }: IProps) => {
  const [selectedAllRows, setSelectedAllRows] = React.useState(false);
  const [selectedRows, setSelectedRows] = React.useState<string[]>([]);
  let keys = Object.keys(body[0]);

  const borderRadiusRow = (item: string, index: number) => {
    return item === "checkbox"
      ? {
          borderRadius: "10px 0 0 10px",
        }
      : {
          borderRadius: keys.length - 1 === index ? "0 10px 10px 0" : "none",
        };
  };

  if (withSelectedChoice) {
    keys = ["checkbox", ...keys];
  }

  const onSelectAllRows = () => {
    if (selectedRows.length > 0) {
      setSelectedRows([]);
    }

    setSelectedAllRows(!selectedAllRows);
  };

  const onSelectedRow = (element: string) => {
    if (selectedAllRows) {
      setSelectedAllRows(false);
    }

    if (selectedRows.includes(element)) {
      return setSelectedRows(selectedRows.filter((item) => item !== element));
    }

    return setSelectedRows([...selectedRows, element]);
  };

  const tableHeadContent = (item: string) => {
    if (item === "checkbox") {
      return (
        <TableCheckboxHeaderContainer>
          <Checkbox isChecked={selectedAllRows} onChange={onSelectAllRows} />
        </TableCheckboxHeaderContainer>
      );
    }

    if (item === "withActions") {
      return "";
    }

    if (item === "Id") {
      return (
        <Row>
          <Information />
          <span style={{ color: "#050505", marginLeft: "8px" }}>ID</span>
        </Row>
      );
    }

    return item;
  };

  const tableBodyContent = (item: string, id: string, parent: any) => {
    if (item === "checkbox") {
      return (
        <TableCheckboxHeaderContainer>
          <Checkbox
            isChecked={
              selectedRows.includes(`selectedRow__${id}`) || selectedAllRows
            }
            onChange={() => onSelectedRow(`selectedRow__${id}`)}
          />
        </TableCheckboxHeaderContainer>
      );
    }

    if (item === "withActions") {
      return (
        <TableActionsContainer style={{ width: "74px" }}>
          <TableActions>
            <Actions />
          </TableActions>
        </TableActionsContainer>
      );
    }

    return <>{parent[item]}</>;
  };

  return (
    <TableContainer role="table" data-testid="TableContainer">
      <TableHeader data-testid="TableHeader" role="rowgroup">
        <TableRow data-testid="TableRow" role="row">
          {keys.map((item: string, i: number) => (
            <TableHead
              data-testid={`TableHead__${i}`}
              style={borderRadiusRow(item, i)}
              key={`${item}__${i}`}
              role="columnheader"
            >
              {tableHeadContent(item)}
            </TableHead>
          ))}
        </TableRow>
      </TableHeader>
      <TableBody data-testid="TableBody" role="rowgroup">
        {body &&
          body.length &&
          body.map((item: any, i: number) => {
            return (
              <TableRow
                data-testid={`TableRow__${i}`}
                key={`${item}__${i}`}
                role="row"
                style={{
                  background:
                    selectedRows.includes(`selectedRow__${item.Id}`) ||
                    selectedAllRows
                      ? "#f0f0f0"
                      : "white",
                }}
              >
                {keys.map((k: string, j: number) => (
                  <TableCell
                    style={{
                      ...borderRadiusRow(k, j),
                    }}
                    key={k}
                    role="columnheader"
                  >
                    {tableBodyContent(k, item.Id, item)}
                  </TableCell>
                ))}
              </TableRow>
            );
          })}
      </TableBody>
    </TableContainer>
  );
};

export default Table;
