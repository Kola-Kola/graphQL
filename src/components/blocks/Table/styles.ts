import styled from "styled-components";

export const TableContainer = styled.table`
  border-collapse: collapse;
  width: 100%;
`;
export const TableHeader = styled.thead`
  background: ${({ theme }) => theme.colors.neutral4Hover};
  height: 64px;
  border-radius: 8px;
`;
export const TableBody = styled.tbody`
  &:before {
    line-height: 24px;
    content: ".";
    display: block;
  }
`;
export const TableRow = styled.tr`
  text-align: left;
  height: 64px;
  transition: all ease-in-out 0.3s;
  cursor: pointer;

  &:hover {
    background: ${({ theme }) => theme.colors.neutral4Hover} !important;
  }

  &:hover {
    td {
      background: ${({ theme }) => theme.colors.neutral4Hover} !important;
    }
  }
`;
export const TableHead = styled.th`
  color: ${({ theme }) => theme.colors.neutral2};
  text-align: left;
  font-family: "Maison Neue", arial;
`;

export const TableCell = styled.td`
  font-size: 16px;
  font-weight: normal;
  color: ${({ theme }) => theme.colors.neutral1};
  font-family: "Maison Neue", arial;
  transition: all ease-in-out 0.3s;
`;

export const TableActionsContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 74px;
`;

export const TableActions = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 40px;
  height: 40px;
  border-radius: 4px;
  transition: all ease-in-out 0.3s;

  &:hover {
    background: ${({ theme }) => theme.colors.neutral4Hover};
  }
`;

export const TableCheckboxHeaderContainer = styled.div`
  display: flex;
  flex-flow: row wrap;
  align-items: center;
  justify-content: center;
  height: 64px;
  width: 48px;
  margin-left: 20px;
`;

export const Row = styled.p`
  display: flex;
  flex-flow: row wrap;
  align-items: center;
`;
