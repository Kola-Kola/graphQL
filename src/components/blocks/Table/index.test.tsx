import { render, screen } from "@testing-library/react"; // (or /dom, /vue, ...)
import Table from "./index";
import renderer from "react-test-renderer";
import Theme from "../../layout/Theme";
import "jest-styled-components";

const TableData = [
  { firstName: "Jonathan", lastName: "IBOR" },
  { firstName: "Jean", lastName: "Dupond" },
];

describe("Table Component", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Theme>
          <Table body={TableData} withSelectedChoice={false} />
        </Theme>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render Table with partial features", () => {
    render(
      <Theme>
        <Table body={TableData} withSelectedChoice={false} />
      </Theme>
    );

    expect(screen.getByTestId("TableContainer")).toBeInTheDocument();
    expect(screen.getByTestId("TableHeader")).toBeInTheDocument();
    expect(screen.getByTestId("TableRow")).toBeInTheDocument();

    expect(screen.getByTestId("TableHead__0")).toBeInTheDocument();
    expect(screen.getByTestId("TableHead__1")).toBeInTheDocument();
    expect(screen.queryByTestId("TableHead__2")).not.toBeInTheDocument();

    expect(screen.getByTestId("TableBody")).toBeInTheDocument();
    expect(screen.getByTestId("TableRow__0")).toBeInTheDocument();
    expect(screen.getByTestId("TableRow__1")).toBeInTheDocument();
    expect(screen.queryByTestId("TableRow__2")).not.toBeInTheDocument();
  });
});
