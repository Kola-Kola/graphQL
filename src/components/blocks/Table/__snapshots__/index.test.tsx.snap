// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component renders correctly 1`] = `
.c0 {
  border-collapse: collapse;
  width: 100%;
}

.c1 {
  background: #F0F0F0;
  height: 64px;
  border-radius: 8px;
}

.c4:before {
  line-height: 24px;
  content: ".";
  display: block;
}

.c2 {
  text-align: left;
  height: 64px;
  -webkit-transition: all ease-in-out 0.3s;
  transition: all ease-in-out 0.3s;
  cursor: pointer;
}

.c2:hover {
  background: #F0F0F0;
}

.c2:hover td {
  background: #F0F0F0 !important;
}

.c3 {
  color: #8C8C8C;
  text-align: left;
  font-family: "Maison Neue",arial;
}

.c5 {
  font-size: 16px;
  font-weight: normal;
  color: #050505;
  font-family: "Maison Neue",arial;
  -webkit-transition: all ease-in-out 0.3s;
  transition: all ease-in-out 0.3s;
}

<table
  className="c0"
  data-testid="TableContainer"
  role="table"
>
  <thead
    className="c1"
    data-testid="TableHeader"
    role="rowgroup"
  >
    <tr
      className="c2"
      data-testid="TableRow"
      role="row"
    >
      <th
        className="c3"
        data-testid="TableHead__0"
        role="columnheader"
        style={
          Object {
            "borderRadius": "none",
          }
        }
      >
        firstName
      </th>
      <th
        className="c3"
        data-testid="TableHead__1"
        role="columnheader"
        style={
          Object {
            "borderRadius": "0 10px 10px 0",
          }
        }
      >
        lastName
      </th>
    </tr>
  </thead>
  <tbody
    className="c4"
    data-testid="TableBody"
    role="rowgroup"
  >
    <tr
      className="c2"
      data-testid="TableRow__0"
      role="row"
    >
      <td
        className="c5"
        role="columnheader"
        style={
          Object {
            "background": "white",
            "borderRadius": "none",
          }
        }
      >
        Jonathan
      </td>
      <td
        className="c5"
        role="columnheader"
        style={
          Object {
            "background": "white",
            "borderRadius": "0 10px 10px 0",
          }
        }
      >
        IBOR
      </td>
    </tr>
    <tr
      className="c2"
      data-testid="TableRow__1"
      role="row"
    >
      <td
        className="c5"
        role="columnheader"
        style={
          Object {
            "background": "white",
            "borderRadius": "none",
          }
        }
      >
        Jean
      </td>
      <td
        className="c5"
        role="columnheader"
        style={
          Object {
            "background": "white",
            "borderRadius": "0 10px 10px 0",
          }
        }
      >
        Dupond
      </td>
    </tr>
  </tbody>
</table>
`;
